mesonpy/__init__.py

```diff
@@ -267,166 +301,34 @@ def _debian_python(self) -> bool:
         except ModuleNotFoundError:
             return False
 
-    @property
-    def python_tag(self) -> str:
-        selected_tag = self._select_abi_tag()
-        if selected_tag and selected_tag.python:
-            return selected_tag.python
-        return 'py3'
-
-    @property
-    def abi_tag(self) -> str:
-        selected_tag = self._select_abi_tag()
-        if selected_tag:
-            return selected_tag.abi
-        return 'none'
-
     @cached_property
-    def platform_tag(self) -> str:
-        if self.is_pure:
-            return 'any'
-        # XXX: Choose the sysconfig platform here and let something like auditwheel
-        #      fix it later if there are system dependencies (eg. replace it with a manylinux tag)
-        platform_ = sysconfig.get_platform()
-        parts = platform_.split('-')
-        if parts[0] == 'macosx':
-            target = os.environ.get('MACOSX_DEPLOYMENT_TARGET')
```

Comment:
We need this code. This breaks `MACOSX_DEPLOYMENT_TARGET` support.

Reply:
AFAIK, none of the other PEP517 backends implement this. I tested this with `cibuildwheel` and it seemed to do the right thing. I guess `cibuildwheel` sets the `_PYTHON_HOST_PLATFORM` or an equivalent environment variable to achieve the same. As highlighted by the bug reports concerning macOS 13, getting the platform tag to agree with what `packaging` thinks it needs to be is going to be a game of whack-a-mole I would really like to avoid if possible.

Reply:
Yes, that's what `cibuildwheel` does: https://github.com/pypa/cibuildwheel/blob/2e6cbe2435ef410156319c98625002ed76e08785/cibuildwheel/macos.py#L219

Reply:
I think that advanced users that require the magic of `MACOS_DEPLOYMENT_TARGET` should also know about `_PYTHON_HOST_PLATFORM`. I would prefer documenting this somewhere than adding special handling here. In general, I think that `meson-python` should behave as close as possible to other Python build tools, to adhere to the principle of less surprise.

Reply:
Yes, but that's private API and meant to be used for cross-builds specifically. We should be migrating away from relying on it, as it could have other unintended behavior changes, or even go away entirely if/when we add better cross-build support in Python.

Reply:
If this is the case, support for `MACOS_DEPLOYMENT_TARGET` should go into `sysconfig` or `packaging` or whatever. `meson-python` tried to have his own code to generate the macOS platform tag, and it is wrong. Letting `packaging` deal with it makes it certain that at least `pip` and `meson-python` will agree on the tags. And makes it one less place that needs to be fixed at the next macOS major version bump, or for the next re-intepretation of the platform tag PEPs.

Reply:
My reason for resisting to implement this is that I don't want to replicate this https://github.com/pypa/packaging/blob/606c71acce93d04e778cdbdea16231b36d6b870f/packaging/tags.py#L346-L431 into `meson-python`. However, at a closer look, that function allows to pass in a version specifier, so problem solved.

Reply:
FWIW, I've liked the way I implemented this in scikit-build-core: 

https://github.com/scikit-build/scikit-build-core/blob/3ab96051f3626f4323f681e5f29afaf029735bc0/src/scikit_build_core/builder/wheel_tag.py#L29-L54
&
https://github.com/scikit-build/scikit-build-core/blob/main/src/scikit_build_core/builder/macos.py

Reply:
Well, we only need the following snippet, to work around the `packaging` bug.

https://github.com/mesonbuild/meson-python/blob/ef67e0e03c3224a8703809ce6eafb8ac5ce352d5/mesonpy/__init__.py#L312-L320

But, arguably, we could even remove it, and require users to se `MACOSX_DEPLOYMENT_TARGET` to a vlaue `packaging` currently accepts. I would like to keep this workaround though, and just remove it if it gives us toruble.

Reply:
@FFY00 I don't understand. If we delegate computing the tag to `packaging` via `packaging.tags.mac_platforms()` we necessarily agree with what `packaging` considers valid. Don't we?

Reply:
See my point above.

> `MACOS_DEPLOYMENT_TARGET` is set to overwrite the target platform, making `sysconfig.get_platform` return it is wrong IMO, and making `packging` doesn't feel right either with the current provided API, as it is primarily used to see if a certain platform is compatible with the running interpreter.

Reply:
We can still use `mac_platforms` if you want, but we should support overwriting that via `MACOS_DEPLOYMENT_TARGET`.

Sorry, I wasn't clear enough above.

Reply:
I think I'm not making my point come across. I was thinking to the equivalent of
```python
try:
    version = tuple(map(int, os.environ['MACOS_DEPLOYMENT_TARGET'].split('.')))
except ValueError:
    version = None
platform = next(packaging.tags.mac_platforms(version))
```
where all and every normalization is delegated to `packaging`.

Reply:
Yes, that makes sense :+1:

Reply:
I would also recommend producing at least a warning if `MACOS_DEPLOYMENT_TARGET` is set lower than `sysconfig.get_platform()`, which is the version Python was compiled with. Maybe an error, as you won't really be producing something that can run on what you think it can run on.

Reply:
> I would also recommend producing at least a warning if `MACOS_DEPLOYMENT_TARGET` is set lower than `sysconfig.get_platform()`, which is the version of macOS that Python was compiled with. Maybe an error, as you won't really be producing something that can run on what you think it can run on. (which is what I was trying to show above)

I don't understand why this is the case. The extensions are not statically linked with the Python libraries.

Reply:
It was due to https://github.com/pypa/setuptools/blob/cd7d6d56a1cf84cafea44f7cf7e357a926821f03/distutils/spawn.py#L62-L70 - which is now removed in favor of a single check - apparently it mattered once, for... 10.3. The configuration values in sysconfig were incorrect if it was compiled for an older macOS than 10.3. The modern version of that is here: https://github.com/pypa/setuptools/blob/ac727af6acd38537cdddc6c98bc9457aa1eb991a/setuptools/_distutils/util.py#L98-L117 (FYI, Python itself hasn't supported < 10.6 in forever and < 10.9 in a while, AFAIK.)

Sysconfig values are a lot less important for us than for distutils, too, so I think I was wrong.